README.md
# CRND Deploy

This is simple script to install and configure production-ready [Odoo](https://www.odoo.com/) instance,
managed by [odoo-helper-scripts](https://katyukha.gitlab.io/odoo-helper-scripts/).

To deploy [Odoo](https://www.odoo.com/) just clone repo to machine and run `sudo crnd-deploy.bash`.
To get install options, just call `sudo crnd-deploy.bash --help` command.

Also, this script supports automatic installation of
[PostgreSQL](https://www.postgresql.org/) and
[Nginx](https://nginx.org/en/) on same machine.

## Features

- Easy installation of production-ready odoo instances.
- Automatic installation and pre-configuration of [Nginx](https://nginx.org/en/) (if needed).
- Automatic installation of [PostgreSQL](https://www.postgresql.org/) (if needed)
- Full power of [odoo-helper-scripts](https://katyukha.gitlab.io/odoo-helper-scripts/) that significantly simplifies the management of the server.


## Yodoo Cockpit

[![Yodoo Cockpit](https://crnd.pro/web/image/18846/banner_2_4_gif_animation_cut.gif)](https://crnd.pro/yodoo-cockpit)

Take a look at [Yodoo Cockpit](https://crnd.pro/yodoo-cockpit) project, and discover the easiest way to manage your odoo installation.
Just short notes about [Yodoo Cockpit](https://crnd.pro/yodoo-cockpit):
- start new production-ready odoo instance in 1-2 minutes.
- add custom addons to your odoo instances in 5-10 minutes.
- out-of-the-box email configuration: just press button and add some records to your DNS, and get a working email
- make your odoo instance available to external world (internet) in 30 seconds (just add single record in your DNS)

If you have any questions, then contact us at [info@crnd.pro](mailto:info@crnd.pro), so we could schedule online-demonstration.

## CRND-Deploy Requirements

Only [Ubuntu](https://ubuntu.com/) 18.04 and 20.04 supported and tested.

May be working on other debian-based linux distributions, but without any warranty.

## Supported Odoo Versions

| Odoo Serie | Support            |
|------------|--------------------|
| 8.0        | ***not tested***   |
| 9.0        | ***not tested***   |
| 10.0       | ***not tested***   |
| 11.0       | :heavy_check_mark: |
| 12.0       | :heavy_check_mark: |
| 13.0       | :heavy_check_mark: |
| 14.0       | :heavy_check_mark: |
| 15.0       | :heavy_check_mark: |
| 16.0       | :heavy_check_mark: |

## Options available

Just call 

```sh
sudo crnd-deploy.bash --help
```

And see help message with all available commands:

```
Usage:

    crnd-deploy.bash [options]    - install odoo

Options:

    --odoo-repo <repo>       - git repository to clone odoo from.
                               default: https://github.com/odoo/odoo
    --odoo-branch <branch>   - odoo branch to clone.
                               default: 12.0
    --odoo-version <version> - odoo version to clone.
                               default: 12.0
    --odoo-user <user>       - name of system user to run odoo with.
                               default: odoo
    --db-host <host>         - database host to be used by odoo.
                               default: localhost
    --db-user <user>         - database user to connect to db with
                               default: odoo
    --db-password <password> - database password to connect to db with
                               default: odoo
    --install-dir <path>     - directory to install odoo in
                               default: /opt/odoo
    --install-mode <mode>    - installation mode. could be: 'git', 'archive'
                               default: git
    --local-postgres         - install local instance of postgresql server
    --proxy-mode             - Set this option if you plan to run odoo
                               behind proxy (nginx, etc)
    --workers <workers>      - number of workers to run.
                               Default: 2
    --local-nginx            - install local nginx and configure it for this
                               odoo instance
    --odoo-helper-dev        - If set then use dev version of odoo-helper
    --install-ua-locales     - If set then install also uk_UA and ru_RU
                               system locales.
    -v|--version             - print version and exit
    -h|--help|help           - show this help message

Suggestion:

    Take a look at [Yodoo Cockpit](https://crnd.pro/yodoo-cockpit) project,
    and discover the easiest way to manage your odoo installation.

    Just short notes about [Yodoo Cockpit](https://crnd.pro/yodoo-cockpit):
        - start new production-ready odoo instance in 1-2 minutes.
        - add custom addons to your odoo instances in 5-10 minutes.
        - out-of-the-box email configuration: just press button and
          add some records to your DNS, and get a working email
        - make your odoo instance available to external world in 30 seconds:
          just add single record in your DNS

    If you have any questions, then contact us at
    [info@crnd.pro](mailto:info@crnd.pro),
    so we could schedule online-demonstration.
```

## Usage

Basically to install [Odoo](https://www.odoo.com/) on new machine you have to do following:

```sh
# Download script from github
git clone https://github.com/crnd-inc/crnd-deploy

# Install odoo
sudo bash crnd-deploy/crnd-deploy.bash --odoo-version 15.0 --local-postgres --local-nginx
```

This command will automatically install and configure [Odoo](https://www.odoo.com/),
[PostgreSQL](https://www.postgresql.org/), [Nginx](https://nginx.org/en/)
on machine, thus you get complete production-ready odoo installation.

After odoo installed, it could be easily managed by [odoo-helper-scripts](https://katyukha.gitlab.io/odoo-helper-scripts/).

For example, git repository with third-party addons could be easily added via command:

```sh
odoo-helper fetch -r <my git repo>
```

Server logs could be viewed with `less` pager via command

```sh
odoo-helper server log
```


## Level up your service quality

Level up your service with our [Helpdesk](https://crnd.pro/solutions/helpdesk) / [Service Desk](https://crnd.pro/solutions/service-desk) / [ITSM](https://crnd.pro/itsm) solution.

Just test it at [yodoo.systems](https://yodoo.systems/saas/templates): choose template you like, and start working.

Test all available features of [Bureaucrat ITSM](https://crnd.pro/itsm) with [this template](https://yodoo.systems/saas/template/bureaucrat-itsm-demo-data-95).


## Bug tracker

Bugs are tracked on [https://crnd.pro/requests](https://crnd.pro/requests>).
In case of trouble, please report there.


## Maintainer

![Center of Research & Development](https://crnd.pro/web/image/3699/300x140/crnd.png)

Our web site is: [crnd.pro](https://crnd.pro/)

This module is maintained by the [Center of Research & Development](https://crnd.pro) company.

We can provide you further Odoo Support, Odoo implementation, Odoo customization, Odoo 3rd Party development and integration software, consulting services (more info available on [our site](https://crnd.pro/our-services)).Our main goal is to provide the best quality product for you. 

For any questions [contact us](mailto:info@crnd.pro>).



CHANGELOG.md
# Releases

## Release 1.2.0

- Change default install mode to archive
- Added options that allows to build custom version of python

## Release 1.1.0

- Updated docs
- Use python3 to run generation of nginx config
- Update default Odoo version to 15.0

## Release 1.0.0

Initial release


gen_nginx.py
#!/usr/bin/env python

##############################################################################
#  CRND Deploy - the simple way to start new production-ready Odoo instance. #
#  Copyright (C) 2020  Center of Research and Development                    #
#                                                                            #
#  This program is free software: you can redistribute it and/or modify      #
#  it under the terms of the GNU General Public License as published by      #
#  the Free Software Foundation, either version 3 of the License, or         #
#  your option) any later version.                                           #
#                                                                            #
#  This program is distributed in the hope that it will be useful,           #
#  but WITHOUT ANY WARRANTY; without even the implied warranty of            #
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the             #
#  GNU General Public License for more details.                              #
#                                                                            #
#  You should have received a copy of the GNU General Public License         #
#  along with this program.  If not, see <https://www.gnu.org/licenses/>.    #
##############################################################################

# Variables:
#   - instance_name
#   - instance_ip
#   - instance_port       (default: 8069)
#   - instance_lp_port    (default: 8072)
#   - frontend_ip
#   - frontend_server_name

NGINX_TEMPLATE = """
upstream crnd_{instance_name} {{
        server {instance_ip}:{instance_port} weight=1 fail_timeout=2000s;
}}

upstream crnd_{instance_name}_longpolling {{
        server {instance_ip}:{instance_lp_port} weight=1 fail_timeout=300s;
}}


# Force SSL (HTTPS)
#server {{
    #listen   {frontend_ip}:80;
    #server_name  {frontend_server_name};

    #location / {{
        #return 301 https://$host$request_uri;
    #}}
#}}

server {{
    listen   {frontend_ip}:80;
    # listen   {frontend_ip}:443 ssl;
    # server_name  {frontend_server_name};

    #-----------------------------------------------------------------------
    access_log  /var/log/nginx/{instance_name}.access.log;
    error_log   /var/log/nginx/{instance_name}.error.log;
    #-----------------------------------------------------------------------

    #-----------------------------------------------------------------------
    # SSL config
    #ssl on;
    #ssl_certificate  /etc/nginx/ssl/server.crt;
    #ssl_certificate_key /etc/nginx/ssl/server.key;
    #-----------------------------------------------------------------------

    #-----------------------------------------------------------------------
    # global params for Odoo backend server section
    client_max_body_size 100m;

    # Proxy global settings
    # increase proxy buffer to handle some OpenERP web requests
    proxy_buffers 16 64k;
    proxy_buffer_size 128k;

    # general proxy settings
    # force timeouts if the backend dies
    proxy_connect_timeout 900s;
    proxy_send_timeout 900s;
    proxy_read_timeout 900s;
    proxy_next_upstream error timeout invalid_header http_500 http_502 http_503;

    # set headers
    proxy_set_header Host $host;
    proxy_set_header X-Forwarded-Proto $scheme;
    proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Forward-For $proxy_add_x_forwarded_for;
    proxy_set_header X-Real-IP $remote_addr;


    # by default, do not forward anything
    proxy_redirect off;
    proxy_buffering off;

    # use gzip for folowing types
    gzip_types text/html text/css text/less text/plain text/xml application/xml application/json application/javascript;
    #-----------------------------------------------------------------------


    location / {{
        add_header Content-Security-Policy "upgrade-insecure-requests";
        proxy_pass http://crnd_{instance_name};
    }}

    # Chat and IM related features support
    location /longpolling {{
        add_header Content-Security-Policy "upgrade-insecure-requests";
        proxy_pass http://crnd_{instance_name}_longpolling;
    }}

    # Restrict access
    location ~* ^/(web/database/|jsonrpc|xmlrpc|web/tests) {{
        # TODO: Restrict external access here
        #    allow trusted_network;
        #    allow trusted_ip;
        #    deny all;
        add_header Content-Security-Policy "upgrade-insecure-requests";
        proxy_pass http://crnd_{instance_name};
    }}

    # cache some static data in memory for 60mins.
    # under heavy load this will preserve the OpenERP Web client a little bit.
    location /web/static/ {{
        proxy_cache_valid 200 60m;
        proxy_buffering    on;
        expires 864000;
        add_header Content-Security-Policy "upgrade-insecure-requests";
        proxy_pass         http://crnd_{instance_name};
    }}
}}
"""

#   - instance_name
#   - instance_ip
#   - instance_port       (default: 8069)
#   - instance_lp_port    (default: 8072)
#   - frontend_ip
#   - frontend_server_name
import argparse
parser = argparse.ArgumentParser(
    description='Simply generates nginx conf and prints it to STDOUT')

template_group = parser.add_argument_group('Template')
template_group.add_argument(
    '--instance-name', required=True,
    help='short name of instance to gen config for')
template_group.add_argument(
    '--instance-ip', default='localhost',
    help='Odoo instance ip')
template_group.add_argument(
    '--instance-port', type=int, default=8069, help='Odoo instance port')
template_group.add_argument(
    '--instance-lp-port', type=int, default=8072,
    help='Odoo instance longpolling port (used for chatter)')
template_group.add_argument(
    '--frontend-ip', default='0.0.0.0',
    help='IP to bind nginx to')
template_group.add_argument(
    '--frontend-server-name', required=True,
    help='nginx servername')
args = parser.parse_args()

generated_conf = NGINX_TEMPLATE.format(**vars(args))
print(generated_conf)



docker/maao-simple/install_odoo_simple.bash
#--------------------------------------------------
# Apply environment configuration
#--------------------------------------------------
ODOO_REPO=${ODOO_REPO:-https://github.com/managment-and-acounting-on-line/maao};
ODOO_BRANCH=${ODOO_BRANCH:-maao-9.0};
ODOO_USER=${ODOO_USER:-odoo};
PROJECT_ROOT_DIR=${ODOO_INSTALL_DIR:-/opt/odoo};
DB_HOST=${ODOO_DB_HOST:-localhost};
DB_PORT=${ODOO_DB_PORT:-5432}
DB_USER=${ODOO_DB_USER:-odoo};
DB_PASSWORD=${ODOO_DB_PASSWORD:-odoo};

#--------------------------------------------------
# Install Odoo
#--------------------------------------------------
echo -e "\n${BLUEC}Installing odoo...${NC}\n";
# import odoo-helper common module, which contains some useful functions
source $(odoo-helper system lib-path common);

# import odoo-helper libs
ohelper_require 'install';

# Do not ask confirmation when installing dependencies
ALWAYS_ANSWER_YES=1;

# Configure default odoo-helper variables
config_default_vars;  # imported from common module
unset VENV_DIR;       # disable vertual environment

# define addons path to be placed in config files
ADDONS_PATH="$ODOO_PATH/openerp/addons,$ODOO_PATH/addons,$ADDONS_DIR";
ODOO_PID_FILE="/var/run/odoo.pid";  # default odoo pid file location

echo "---Create project tree---"
install_create_project_dir_tree;   # imported from 'install' module
# install_system_prerequirements;    # imported from 'install' module


# install_and_configure_postgresql;   # imported from 'install' module

if [ ! -d $ODOO_PATH ]; then
    echo "--- Clonning odoo ---";
    install_clone_odoo;   # imported from 'install' module
else
    echo "--- Odoo seems to be installed ---";
fi

echo "--- installing system dependencies ---"
install_sys_deps;   # imported from 'install' module
#install_python_prerequirements 1;   # imported from 'install' module

echo "--- installing odoo ---"
# Run setup.py with gevent workaround applied.
odoo_run_setup_py;  # imported from 'install' module

# generate odoo config file
declare -A ODOO_CONF_OPTIONS;
ODOO_CONF_OPTIONS[addons_path]="$ADDONS_PATH";
ODOO_CONF_OPTIONS[admin_passwd]="${ODOO_ADMIN_PASS:-$(random_string 32)}";
ODOO_CONF_OPTIONS[data_dir]="$DATA_DIR";
ODOO_CONF_OPTIONS[logfile]="";  # No log file for docker odoo instance
ODOO_CONF_OPTIONS[db_host]="$DB_HOST";
ODOO_CONF_OPTIONS[db_port]="$DB_PORT";
ODOO_CONF_OPTIONS[db_user]="$DB_USER";
ODOO_CONF_OPTIONS[db_password]="$DB_PASSWORD";
ODOO_CONF_OPTIONS[pidfile]="";
install_generate_odoo_conf $ODOO_CONF_FILE;   # imported from 'install' module

# Write odoo-helper project config
echo "#---ODOO-INSTANCE-CONFIG---" >> /etc/$CONF_FILE_NAME;
echo "`print_helper_config`" >> /etc/$CONF_FILE_NAME;

echo -e "\n${GREENC}Odoo installed!${NC}\n";



crnd-deploy.bash
#!/bin/bash

# CRND Deploy - the simple way to start new production-ready Odoo instance.
# Copyright (C) 2020  Center of Research and Development
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.


# WARN: Must be ran under SUDO

# NOTE: Automaticaly installs odoo-helper-scripts if not installed yet

# Supports passing parametrs as environment variables and as arguments to script
# Environment vars and default values:
#   ODOO_USER=odoo
#   ODOO_INSTALL_DIR=/opt/odoo
#   ODOO_DB_HOST=localhost
#   ODOO_DB_USER=odoo
#   ODOO_DB_PASSWORD=odoo
#   ODOO_REPO=https://github.com/odoo/odoo
#   ODOO_BRANCH=15.0
#   ODOO_VERSION=15.0
#   ODOO_WORKERS=2
#
# Also some configuration could be passed as command line args:
#   sudo bash crnd-deploy.bash <db_host> <db_user> <db_pass>
# 

#--------------------------------------------------
# Script params
#--------------------------------------------------
SCRIPT=$0;
SCRIPT_NAME=$(basename $SCRIPT);
SCRIPT_DIR=$(dirname $SCRIPT);
SCRIPT_PATH=$(readlink -f $SCRIPT);
NGIX_CONF_GEN="$SCRIPT_DIR/gen_nginx.py";

WORKDIR=`pwd`;

#--------------------------------------------------
# Version
#--------------------------------------------------
CRND_DEPLOY_VERSION="1.1.1"

#--------------------------------------------------
# Defaults
#--------------------------------------------------
DEFAULT_ODOO_BRANCH=15.0
DEFAULT_ODOO_VERSION=15.0
#--------------------------------------------------
# Parse environment variables
#--------------------------------------------------
ODOO_REPO=${ODOO_REPO:-https://github.com/odoo/odoo};
ODOO_BRANCH=${ODOO_BRANCH:-$DEFAULT_ODOO_BRANCH};
ODOO_VERSION=${ODOO_VERSION:-$DEFAULT_ODOO_VERSION};
ODOO_USER=${ODOO_USER:-odoo};
ODOO_WORKERS=${ODOO_WORKERS:-2};
PROJECT_ROOT_DIR=${ODOO_INSTALL_DIR:-/opt/odoo};
DB_HOST=${ODOO_DB_HOST:-localhost};
DB_USER=${ODOO_DB_USER:-odoo};
DB_PASSWORD=${ODOO_DB_PASSWORD:-odoo};
INSTALL_MODE=${INSTALL_MODE:-archive};


#--------------------------------------------------
# Define color variables
#--------------------------------------------------
NC='\e[0m';
REDC='\e[31m';
GREENC='\e[32m';
YELLOWC='\e[33m';
BLUEC='\e[34m';
LBLUEC='\e[94m';


if [[ $UID != 0 ]]; then
    echo -e "${REDC}ERROR${NC}";
    echo -e "${YELLOWC}Please run this script as root or with sudo:${NC}"
    echo -e "${BLUEC}sudo $0 $* ${NC}"
    exit 1
fi


#--------------------------------------------------
# FN: Print usage
#--------------------------------------------------
function print_usage {

    echo "
Usage:

    crnd-deploy.bash [options]    - install odoo

Options:

    --odoo-repo <repo>         - git repository to clone odoo from.
                                 default: $ODOO_REPO
    --odoo-branch <branch>     - odoo branch to clone.
                                 default: $ODOO_BRANCH
    --odoo-version <version>   - odoo version to clone.
                                 default: $ODOO_VERSION
    --odoo-user <user>         - name of system user to run odoo with.
                                 default: $ODOO_USER
    --build-python <ver>       - build custom python for specified version
                                 'auto' could be specified to automatically guess correct version.
    --build-python-if-needed   - Automatically detect if it is necessary to build custom python.
    --build-python-optimize    - Apply '--enable-optimizations' for python build.
                                 This could take a while.
    --build-python-sqlite3     - Apply  --enable-loadable-sqlite-extensions
                                 when building python.
    --node-version <ver>       - Version of node.js to be installed. By default lts.
    --db-host <host>           - database host to be used by odoo.
                                 default: $DB_HOST
    --db-user <user>           - database user to connect to db with
                                 default: $DB_USER
    --db-password <password>   - database password to connect to db with
                                 default: $DB_PASSWORD
    --install-dir <path>       - directory to install odoo in
                                 default: $PROJECT_ROOT_DIR
    --install-mode <mode>      - installation mode. could be: 'git', 'archive'
                                 default: $INSTALL_MODE
    --local-postgres           - install local instance of postgresql server
    --proxy-mode               - Set this option if you plan to run odoo
                                 behind proxy (nginx, etc)
    --workers <workers>        - number of workers to run.
                                 Default: $ODOO_WORKERS
    --local-nginx              - install local nginx and configure it for this
                                 odoo instance
    --odoo-helper-dev          - If set then use dev version of odoo-helper
    --install-ua-locales       - If set then install also uk_UA and ru_RU
                                 system locales.
    -v|--version               - print version and exit
    -h|--help|help             - show this help message

Suggestion:

    Take a look at [Yodoo Cockpit](https://crnd.pro/yodoo-cockpit) project,
    and discover the easiest way to manage your odoo installation.

    Just short notes about [Yodoo Cockpit](https://crnd.pro/yodoo-cockpit):
        - start new production-ready odoo instance in 1-2 minutes.
        - add custom addons to your odoo instances in 5-10 minutes.
        - out-of-the-box email configuration: just press button and
          add some records to your DNS, and get a working email
        - make your odoo instance available to external world in 30 seconds:
          just add single record in your DNS

    If you have any questions, then contact us at
    [info@crnd.pro](mailto:info@crnd.pro),
    so we could schedule online-demonstration.

---
Version: ${CRND_DEPLOY_VERSION}
";
}

#--------------------------------------------------
# Parse command line
#--------------------------------------------------
while [[ $# -gt 0 ]]
do
    key="$1";
    case $key in
        --odoo-repo)
            ODOO_REPO=$2;
            shift;
        ;;
        --odoo-branch)
            ODOO_BRANCH=$2;
            shift;
        ;;
        --odoo-version)
            ODOO_VERSION=$2;
            shift;

            if [ "$ODOO_VERSION" != "$DEFAULT_ODOO_VERSION" ] && [ "$ODOO_BRANCH" == "$DEFAULT_ODOO_BRANCH" ]; then
                ODOO_BRANCH=$ODOO_VERSION;
            fi
        ;;
        --odoo-user)
            ODOO_USER=$2;
            shift;
        ;;
        --build-python)
            ODOO_BUILD_PYTHON_VERSION=$2;
            shift;
        ;;
        --build-python-if-needed)
            ODOO_BUILD_PYTHON_VERSION=auto;
        ;;
        --build-python-optimize)
            ODOO_BUILD_PYTHON_OPTIMIZE=1;
        ;;
        --build-python-sqlite3)
            ODOO_BUILD_PYTHON_SQLITE3=1;
        ;;
        --node-version)
            ODOO_INSTALL_NODE_VERSION=$2;
            shift;
        ;;
        --db-host)
            DB_HOST=$2;
            shift;
        ;;
        --db-user)
            DB_USER=$2;
            shift;
        ;;
        --db-password)
            DB_PASSWORD=$2;
            shift;
        ;;
        --install-dir)
            PROJECT_ROOT_DIR=$2;
            shift;
        ;;
        --install-mode)
            if [ "$2" != "git" ] && [ "$2" != "archive" ]; then
                echo "ERROR: Wrong install mode specified: $2"
                exit 1;
            fi
            INSTALL_MODE=$2;
            shift;
        ;;
        --workers)
            ODOO_WORKERS=$2;
            shift;
        ;;
        --proxy-mode)
            PROXY_MODE=1;
        ;;
        --local-postgres)
            # Generate random password for database
            DB_HOST="localhost";
            DB_PASSWORD="$(< /dev/urandom tr -dc A-Za-z0-9 | head -c 32)";
            INSTALL_LOCAL_POSTGRES=1;
        ;;
        --local-nginx)
            INSTALL_LOCAL_NGINX=1;
            PROXY_MODE=1;
        ;;
        --odoo-helper-dev)
            USE_DEV_VERSION_OF_ODOO_HElPER=1;
        ;;
        --install-ua-locales)
            CRND_DEPLOY_INSTALL_UA_LOCALES=1;
        ;;
        -v|--version)
            echo "$CRND_DEPLOY_VERSION";
            exit 0;
        ;;
        -h|--help|help)
            print_usage;
            exit 0;
        ;;
        *)
            echo "Unknown option global option /command $key";
            echo "Use --help option to get info about available options.";
            exit 1;
        ;;
    esac;
    shift;
done;
#--------------------------------------------------


set -e;   # fail on errors

#--------------------------------------------------
# Update Server and Install Dependencies
#--------------------------------------------------
echo -e "\n${BLUEC}Update Server...${NC}\n";
apt-get update -qq;
apt-get upgrade -qq -y;
echo -e "\n${BLUEC}Installing basic dependencies...${NC}\n";
apt-get install -qqq -y \
    sudo wget locales;

#--------------------------------------------------
# Generate locales
#--------------------------------------------------
echo -e "\n${BLUEC}Update locales...${NC}\n";
sed -i -r "s@# en_US.UTF-8(.*)@en_US.UTF-8\1@" /etc/locale.gen;
sed -i -r "s@# en_GB.UTF-8(.*)@en_GB.UTF-8\1@" /etc/locale.gen;
locale-gen en_US.UTF-8;
locale-gen en_GB.UTF-8;

if [ -n "$CRND_DEPLOY_INSTALL_UA_LOCALES" ]; then
    sed -i -r "s@# ru_UA.UTF-8(.*)@ru_UA.UTF-8\1@" /etc/locale.gen;
    sed -i -r "s@# uk_UA.UTF-8(.*)@uk_UA.UTF-8\1@" /etc/locale.gen;
    locale-gen ru_UA.UTF-8;
    locale-gen uk_UA.UTF-8;
fi

update-locale LANG="en_US.UTF-8";
update-locale LANGUAGE="en_US:en";

#--------------------------------------------------
# Ensure odoo-helper installed
#--------------------------------------------------
if ! command -v odoo-helper >/dev/null 2>&1; then
    echo -e "${BLUEC}Odoo-helper not installed! installing...${NC}";
    if ! wget -q -T 2 -O /tmp/odoo-helper-install.bash \
            https://raw.githubusercontent.com/katyukha/odoo-helper-scripts/master/install-system.bash; then
        echo "${REDC}ERROR${NC}: Cannot download odoo-helper-scripts installer from github. Check your network connection.";
        exit 1;
    fi

    # install latest version of odoo-helper scripts
    if [ -z "$USE_DEV_VERSION_OF_ODOO_HElPER" ]; then
        ALWAYS_ANSWER_YES=1 bash /tmp/odoo-helper-install.bash master;
    else
        ALWAYS_ANSWER_YES=1 bash /tmp/odoo-helper-install.bash dev;
    fi

    # Print odoo-helper version
    odoo-helper --version;
fi

# Install odoo pre-requirements
odoo-helper install pre-requirements -y;
odoo-helper install sys-deps -y --branch "$ODOO_BRANCH" "$ODOO_VERSION";

if [ ! -z $INSTALL_LOCAL_POSTGRES ]; then
    odoo-helper install postgres;

    if ! odoo-helper exec postgres_test_connection; then
        echo -e "${YELLOWC}WARNING${NC}: it seams postgres not started, so start it before creating postgres user.";

        # It seams we ran inside docker container, so start postgres server before user creation
        /etc/init.d/postgresql start;
        odoo-helper postgres user-create $DB_USER $DB_PASSWORD;
        /etc/init.d/postgresql stop;
    else
        odoo-helper postgres user-create $DB_USER $DB_PASSWORD;
    fi
fi

#--------------------------------------------------
# Install Odoo
#--------------------------------------------------
echo -e "\n${BLUEC}Installing odoo...${NC}\n";
# import odoo-helper common module, which contains some useful functions
source $(odoo-helper system lib-path common);

# import odoo-helper libs
ohelper_require 'install';
ohelper_require 'config';

# Do not ask confirmation when installing dependencies
ALWAYS_ANSWER_YES=1;

# Configure default odoo-helper variables
config_set_defaults;  # imported from common module

# define addons path to be placed in config files
ADDONS_PATH="$ODOO_PATH/openerp/addons,$ODOO_PATH/odoo/addons,$ODOO_PATH/addons,$ADDONS_DIR";
INIT_SCRIPT="/etc/init.d/odoo";
ODOO_PID_FILE="/var/run/odoo.pid";  # default odoo pid file location

install_create_project_dir_tree;   # imported from 'install' module

if [ ! -d $ODOO_PATH ]; then
    if [ "$INSTALL_MODE" == "git" ]; then
        install_clone_odoo;   # imported from 'install' module
    elif [ "$INSTALL_MODE" == "archive" ]; then
        install_download_odoo;
    else
        echo -e "${REDC}ERROR:${NC} wrong install mode specified: '$INSTALL_MODE'!";
    fi
fi

# install odoo itself
install_odoo_install;  # imported from 'install' module

# generate odoo config file
declare -A ODOO_CONF_OPTIONS;
ODOO_CONF_OPTIONS[addons_path]="$ADDONS_PATH";
ODOO_CONF_OPTIONS[admin_passwd]="$(random_string 32)";
ODOO_CONF_OPTIONS[data_dir]="$DATA_DIR";
ODOO_CONF_OPTIONS[logfile]="$LOG_FILE";
ODOO_CONF_OPTIONS[db_host]="$DB_HOST";
ODOO_CONF_OPTIONS[db_port]="False";
ODOO_CONF_OPTIONS[db_user]="$DB_USER";
ODOO_CONF_OPTIONS[db_password]="$DB_PASSWORD";
ODOO_CONF_OPTIONS[workers]=$ODOO_WORKERS;

# pid file will be managed by init script, not odoo itself
ODOO_CONF_OPTIONS[pidfile]="None";

if [ ! -z $PROXY_MODE ]; then
    ODOO_CONF_OPTIONS[proxy_mode]="True";
fi

install_generate_odoo_conf $ODOO_CONF_FILE;   # imported from 'install' module

# Write odoo-helper project config
echo "#---ODOO-INSTANCE-CONFIG---" >> /etc/$CONF_FILE_NAME;
echo "`config_print`" >> /etc/$CONF_FILE_NAME;

# this will make odoo helper scripts to run odoo with specified user
# (via sudo call)
echo "SERVER_RUN_USER=$ODOO_USER;" >> /etc/$CONF_FILE_NAME;

#--------------------------------------------------
# Fix odoo 9/10 addons path compatability
#--------------------------------------------------
if [ ! -d $ODOO_PATH/openerp/addons ]; then
    sudo mkdir -p $ODOO_PATH/openerp/addons;
fi
if [ ! -d $ODOO_PATH/odoo/addons ]; then
    sudo mkdir -p $ODOO_PATH/odoo/addons;
fi

#--------------------------------------------------
# Create Odoo User
#--------------------------------------------------
if ! getent passwd $ODOO_USER  > /dev/null; then
    echo -e "\n${BLUEC}Createing Odoo user: $ODOO_USER ${NC}\n";
    sudo adduser --system --no-create-home --home $PROJECT_ROOT_DIR \
        --quiet --group $ODOO_USER;
else
    echo -e "\n${YELLOWC}Odoo user already exists, using it.${NC}\n";
fi

#--------------------------------------------------
# Create Init Script
#--------------------------------------------------
echo -e "\n${BLUEC}Creating init script${NC}\n";
sudo cp $ODOO_PATH/debian/init /etc/init.d/odoo
sudo chmod a+x /etc/init.d/odoo
sed -i -r "s@DAEMON=(.*)@DAEMON=$(get_server_script)@" /etc/init.d/odoo;
sed -i -r "s@CONFIG=(.*)@CONFIG=$ODOO_CONF_FILE@" /etc/init.d/odoo;
sed -i -r "s@LOGFILE=(.*)@LOGFILE=$LOG_FILE@" /etc/init.d/odoo;
sed -i -r "s@USER=(.*)@USER=$ODOO_USER@" /etc/init.d/odoo;
sed -i -r "s@PIDFILE=(.*)@PIDFILE=$ODOO_PID_FILE@" /etc/init.d/odoo;
sed -i -r "s@PATH=(.*)@PATH=\1:$VENV_DIR/bin@" /etc/init.d/odoo;
sudo update-rc.d odoo defaults

# Configuration file
sudo chown root:$ODOO_USER $ODOO_CONF_FILE;
sudo chmod 0640 $ODOO_CONF_FILE;

# Log
sudo chown $ODOO_USER:$ODOO_USER $LOG_DIR;
sudo chmod 0750 $LOG_DIR

# Data dir
sudo chown $ODOO_USER:$ODOO_USER $DATA_DIR;

# Odoo root dir
sudo chown $ODOO_USER:$ODOO_USER $PROJECT_ROOT_DIR;

#--------------------------------------------------
# Configure logrotate
#--------------------------------------------------
echo -e "\n${BLUEC}Configuring logrotate${NC}\n";
sudo cat > /etc/logrotate.d/odoo << EOF
$LOG_DIR/*.log {
    copytruncate
    missingok
    notifempty
}
EOF

echo -e "\n${GREENC}Odoo installed!${NC}\n";

if [ ! -z $INSTALL_LOCAL_NGINX ]; then
    echo -e "${BLUEC}Installing and configuring local nginx..,${NC}";
    NGINX_CONF_PATH="/etc/nginx/sites-available/$(hostname).conf";
    sudo apt-get install -qqq -y --no-install-recommends nginx;
    odoo-helper python "$NGIX_CONF_GEN" \
        --instance-name="$(hostname -s)" \
        --frontend-server-name="$(hostname)" > $NGINX_CONF_PATH;
    echo -e "${GREENC}Nginx seems to be installed and default config is generated. ";
    echo -e "Look at $NGINX_CONF_PATH for nginx config.${NC}";
fi
